### ISSUE TITLE
Disallow commas in annotation fields

### ISSUE THREAD

**Issue by sidneymbell** [chanzuckerberg/cellxgene#1029]
On behalf of computational biologists everywhere, **please** either:

1 - Disallow commas in category & label names [preferred] 
OR
2 - Disallow `\t` characters in category & label names and switch output to `.tsv`

Otherwise parsing these files will be an absolute bear and everyone will be sad :(
![image](https://user-images.githubusercontent.com/12618847/68440042-aa20ba00-017e-11ea-822e-b75e31dc039b.png)

**Comment by bkmartinjr** [comment]
@colinmegill - happy to take this.  Do you agree this is essentially an input validation problem?

@sidneymbell - I think we need to define legal syntax more holistically, including the presumed character encoding (eg, are we always ASCII, UTF-8, or?).   There are other dangerous characters, ie, likely to cause problems for downstream use (ie, require quoting, etc), including multi-byte encoded characters (non-ASCII codepoints).  Comma is actually relatively simple, as the CSV syntax require it to be escaped.

If we want to make label files super safe, I propose legal labels are exclusively printable ASCII code points 33 to 126, _excluding_:   space, comma, double-quote, single-quote, back-tick

In other words, they are:    [0-9][a-z][A-Z] plus most punctuation (excluding:  [,'"`] and space)

This is not a very internationally friendly solution, but is practical given that ".csv" does not include any means of specifying *which* character set it is encoded in.  

The other alternative would be to explore the full UTF-8 unicode character set, but that puts the burden on cbio's to handle multi-byte encoding.   Sticking with ASCII makes it much simpler and less error prone, but at the cost of non-English language support.

**Comment by sidneymbell** [comment]
ASCII seems reasonable. I think people will want the space, however, and it
doesn't interfere with csv parsing.


Sidney Bell, PhD
Computational Biology
Chan Zuckerberg Initiative

she, her | twitter <https://twitter.com/sidneymbell> | github
<https://github.com/sidneymbell> | calendar
<https://calendly.com/sidneymbell>

**Comment by colinmegill** [comment]
@bkmartinjr happy to have you handle it, yes input validation though I don't know if this is the only place we'll want to check to be double sure?

**Comment by bkmartinjr** [comment]
@colinmegill - I am happy to handle it.

@sidneymbell - Please keep in mind that including space will add some complexity in processing, as the user must be careful to remove them.  It also makes them visually similar, which will be a problem for users.

Example:
>label, label,  label,label ,
are four unique labels, which will be visually identical in the UI.

I think we should exclude space, or at a minimum, exclude leading, trailing and repeating spaces (at which point, why not just exclude spaces).

**Comment by bkmartinjr** [comment]
One more detail on this besides the legal character codes.   There are two places we can enforce these rules on legal category and label names:
* when the user enters a new name in the UI
* when we load a CSV of labels from the CLI

I propose that we do the following:
* the UI can _display_ and correctly handle any Unicode string for both category & label names (it already does this)
* the UI will only allow the user to specify a new name that conforms to the above rules
* the CLI will load anything that fits in the CSV, and will not enforce rules

This lets us load and display data from other sources, who do not follow the same 'rules', but still enforce our opinionated rules of 'good behavior' when edits originate in our UI.

^ @sidneymbell - please confirm.  This change is forcing an opinion of "good vs bad" behavior on naming, so I want to make sure we agree where it should do so.

**Comment by bkmartinjr** [comment]
@sidneymbell @liaprins-czi - I have been thinking about what is good practice for category/label names.  I would like to suggest we simply use the same convention that is _typically_ used with pandas, ie, a legal Python variable name.   These are alpha-numeric, plus underscore, and are case sensitive.

In other words, punctuation that can cause parsing problems is removed, and spaces which cause visual ambiguity are removed.

legal:   `foo_bar`, `_bar_foo`, `camelCase`, `Snake_case`, `etc`
not-legal: `foo     bar`, `wow!`, `some,stuff`, `...`

Lets chat f2f and nail this down.  I think that Sidney's precautionary approach is sensible - I'm just taking it to the (IMHO) logical conclusion.
